### PR TITLE
ARROW-7349: [C++] Fix the bug of parsing string hex values

### PR DESCRIPTION
**Pull request by liyafan82**
std::lower_bound returns the end of the search range, when failing to find a match.

The end of the search range is one position after the last valid position. So the value in this position is undefined, and we should not reference the value here to compare it with the target value.

### CODE REVIEW

File: cpp/src/arrow/util/string_test.cc [hunk 1]
```diff
@@ -35,5 +37,12 @@ TEST(Trim, Basics) {
   }
 }
 
+TEST(ParseHexValue, Invalid) {
+  std::string input = "xy";
+  uint8_t output;
+  
+  ASSERT_RAISES(Invalid, ParseHexValue(input.c_str(), &output));
```
Comment (pitrou) [inline cpp/src/arrow/util/string_test.cc]: Please also test with an invalid string such as `"@?"`.
Also, while you're at it, add a couple tests for valid values.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/util/string_test.cc (hunk 1 above)
Comment (liyafan82) [reply cpp/src/arrow/util/string_test.cc]: @pitrou Thanks a lot for your valuable comments.
My previous fix was not complete. I have fixed it, and added both positive and negative test cases.
Please take a look. Thank you in advance.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @liyafan82 ! I will merge when green.

**Comment by liyafan82** [comment]
> Thank you @liyafan82 ! I will merge when green.

@pitrou Thank you so much for your effort.

**Comment by wesm** [comment]
Seems this maven error is occurring a lot

```
[ERROR] Failed to execute goal on project arrow-orc: Could not resolve dependencies for project org.apache.arrow.orc:arrow-orc:jar:1.0.0-SNAPSHOT: Failed to collect dependencies at org.apache.orc:orc-core:jar:1.5.5: Failed to read artifact descriptor for org.apache.orc:orc-core:jar:1.5.5: Could not transfer artifact org.apache.orc:orc-core:pom:1.5.5 from/to central (https://repo.maven.apache.org/maven2): repo.maven.apache.org: Temporary failure in name resolution: Unknown host repo.maven.apache.org: Temporary failure in name resolution -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :arrow-orc
##[error]Process completed with exit code 1.
```

Merging

**Pull request closed without merging**